freqtrade/exchange/__init__.py

```diff
@@ -329,6 +358,90 @@ def get_ticker(self, pair: str, refresh: Optional[bool] = True) -> dict:
+    async def _async_get_history(self, pair: str,
+                                 tick_interval: str,
+                                 since_ms: int) -> List:
+        # Assume exchange returns 500 candles
+        _LIMIT = 500
+
+        one_call = constants.TICKER_INTERVAL_MINUTES[tick_interval] * 60 * _LIMIT * 1000
+        logger.debug("one_call: %s", one_call)
+        input_coroutines = [self._async_get_candle_history(
+            pair, tick_interval, since) for since in
+            range(since_ms, arrow.utcnow().timestamp * 1000, one_call)]
+        tickers = await asyncio.gather(*input_coroutines, return_exceptions=True)
+
+        # Combine tickers
+        data: List = []
+        for tick in tickers:
+            if tick[0] == pair:
+                data.extend(tick[1])
+        logger.info("downloaded %s with length %s.", pair, len(data))
+        return data
+
+    async def async_get_candles_history(self, pairs: List[str],
+                                        tick_interval: str) -> List[Tuple[str, List]]:
+        """Download ohlcv history for pair-list asyncronously """
+        input_coroutines = [self._async_get_candle_history(
+            symbol, tick_interval) for symbol in pairs]
+        tickers = await asyncio.gather(*input_coroutines, return_exceptions=True)
+        return tickers
+
+    async def _async_get_candle_history(self, pair: str, tick_interval: str,
+                                        since_ms: Optional[int] = None) -> Tuple[str, List]:
+        try:
+            # fetch ohlcv asynchronously
+            logger.debug("fetching %s since %s ...", pair, since_ms)
+
+            # Calculating ticker interval in second
+            interval_in_sec = constants.TICKER_INTERVAL_MINUTES[tick_interval] * 60
+
+            # If (last update time) + (interval in second) is greater or equal than now
+            # that means we don't have to hit the API as there is no new candle
+            # so we fetch it from local cache
+            if (not since_ms and
+                    self._pairs_last_refresh_time.get(pair, 0) + interval_in_sec >=
+                    arrow.utcnow().timestamp):
+                data = self._cached_klines[pair]
+                logger.debug("Using cached klines data for %s ...", pair)
+            else:
+                data = await self._api_async.fetch_ohlcv(pair, timeframe=tick_interval,
+                                                         since=since_ms)
+
+            # Because some exchange sort Tickers ASC and other DESC.
+            # Ex: Bittrex returns a list of tickers ASC (oldest first, newest last)
+            # when GDAX returns a list of tickers DESC (newest first, oldest last)
+            data = sorted(data, key=lambda x: x[0])
+
+            # keeping last candle time as last refreshed time of the pair
+            self._pairs_last_refresh_time[pair] = data[-1][0] // 1000
+
+            # keeping candles in cache
+            self._cached_klines[pair] = data
```

Comment:
No, I didn't want to touch that one. but maybe we should use ```self.exchange._cached_klines``` in ```freqtradebot.py```? also _klines_last_fetched_time is to be removed.

Reply:
oops _cached_klines is protected in exchange ...

Reply:
what's your recommandation ? I can make it public or do a getter/setter for it to be able to access from freqtradebot.py

Reply:
keeping it in exchange seems to make sense, it's 100% exchange/pair related.

it's protected because it starts with a `_` - which is a convention, not a rule (so you can technically still access it - but shouldn't).

it suits a property well, i agree, but without _cached probably ... ?
so you can access with ``self.exchange.klines['pair']` ... ? 

it's just a idea though ... i'd leave this decision up to you tbh